### PR TITLE
[WIP] improve builds on device

### PR DESCRIPTION
**Pull request by Grimler91**
With this most packages should be safe for builds on device. Instead of installing the package to $PREFIX we use proot to install to a fake $PREFIX in $TERMUX_PKG_MASSAGEDIR, and then generate a deb just as when cross-compiling. 

Packages with host builds needs some special treatment, packages overriding termux_step_make_install needs to be modified to use termux-build-chroot in the install step and all termux_step_post_make_install steps needs to be reviewed, but this works for a few packages I tested (libandroid-support, emacs, tor, binutils).

For libandroid-support and emacs we need to change change the step_make_install function and the termux_step_post_configure steps, respectively.

### CODE REVIEW

**Comment by buttaface** [comment]
Interesting, good to see someone else working on this. I'm not sure why host builds need to be changed, but I haven't messed with proot much. One great result from this approach would be that all the packages that are currently disabled for removing files from the install directory could be enabled for on-device builds.

Could this be used to do cross-builds too, with some modification to set up a fresh chroot for those architectures? I wonder if that wouldn't be the best approach for the host architecture too, so that you don't have to bother with mapping the currently installed packages to the build chroot either, ie make anyone building a package on-device build or install all the package dependencies they need from scratch.

**Comment by Grimler91** [comment]
> I'm not sure why host builds need to be changed, but I haven't messed with proot much. 

I am mostly thinking that it is not needed, and if we do not run termux_step_host_build then most termux_step_{pre,post}_configure steps will fail for these packages (since $TERMUX_PKG_HOSTBUILD_DIR does not exist). 

> One great result from this approach would be that all the packages that are currently disabled for removing files from the install directory could be enabled for on-device builds.

Exactly, at least the majority of them should be safe with these changes.

> Could this be used to do cross-builds too, with some modification to set up a fresh chroot for those architectures? I wonder if that wouldn't be the best approach for the host architecture too, so that you don't have to bother with mapping the currently installed packages to the build chroot either, ie make anyone building a package on-device build or install all the package dependencies they need from scratch.

Allowing for cross-compilation on device would be neat, and for that we would need the host build step anyways, so maybe it is best to always run it even if it is not necessary.

All the dependencies need to be made/downloaded for the other arch though as you said, so we need to create a prooted PREFIX (somewhere else than in TERMUX_PKG_MASSAGEDIR) and use this for not only the make install step but also for the configure and make steps. This would be a slightly larger set of changes than this make install hack, and I think I want to merge this future and consider that further improvement/future work

File: scripts/termux-build-chroot.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	# # Set $TERMUX_PKG_BUILDDIR as current directory:
+	# ARGS="$ARGS --cwd=$TERMUX_PKG_BUILDDIR"
+
+	# Root of the file system:
+	ARGS="$ARGS -r $(realpath $PREFIX/..)"
+
+	# Shell to execute:
+	PROGRAM=bash
+	if [ -x $HOME/.termux/shell ]; then
+		PROGRAM=`readlink -f $HOME/.termux/shell`
```
Comment (xeffyr) [inline scripts/termux-build-chroot.sh]: Any point of using something other than `bash` for build chroot ?

Comment (Grimler91) [reply scripts/termux-build-chroot.sh]: No, this is a leftover from the original termux-chroot script, fixed now

**Comment by Grimler91** [comment]
Changed the approach here. Instead of using a chroot script we just set `DESTDIR=` our massagedir, as done on for example archlinux when compiling packages. We then do not need to try to get cargo/golang/others working in a chroot, which showed to be difficult. It has some shortcomings though:

* A lot of build scripts need to be updated. It works mostly out of the box for autoconf projects, but if we have custom `make install` steps for example, they need to be changed. 
* It does not make it easier to cross-compile packages for other arches as discussed earlier in this thread, for that we need some sort of chroot script probably.

I have updated some packages here as a sort of proof-of-concept, but the heavy lifting, with build.sh editing and testing, is still needed.

**Comment by Grimler91** [comment]
A note to myself so I do not forget it: the packages with hostbuild steps needs to be fixed to work on device. 

Idea is to change so that the host-built tools are built for ubuntu ("amd64"), aarch64, arm, i686, x86_64, and saved as debs in a new repo. When building the hostbuild deb for the same arch is downloaded, if possible, instead of being built every time. These debs can only be built from the docker image, on device only the one for the same arch (aarch64/arm/i686/x86_64) can be built (without extra effort). This feature will be added from another branch and another PR.

**Comment by boypeet** [comment]
> With this most packages should be safe for builds on device. Instead of installing the package to $PREFIX we use proot to install to a fake $PREFIX in $TERMUX_PKG_MASSAGEDIR, and then generate a deb just as when cross-compiling.
> 
> Packages with host builds needs some special treatment, packages overriding termux_step_make_install needs to be modified to use termux-build-chroot in the install step and all termux_step_post_make_install steps needs to be reviewed, but this works for a few packages I tested (libandroid-support, emacs, tor, binutils).
> 
> For libandroid-support and emacs we need to change change the step_make_install function and the termux_step_post_configure steps, respectively.

**Comment by Sukesharch** [comment]
@sukeshsukesh7

**Comment by thunder-coding** [comment]
Suggestion: can you also work on to cache `go`, `cargo` and other dependencies, most mobile devices don't have internet fast enough to download them all in a go, that too multiplayer times not so fast processors.

**Comment by Grimler91** [comment]
> Suggestion: can you also work on to cache `go`, `cargo` and other dependencies, most mobile devices don't have internet fast enough to download them all in a go, and nor so fast processors.

Are you talking about the various host tools that are installed with the [setup scripts](https://github.com/termux/termux-packages/tree/master/scripts/build/setup)? On device they are just installed through apt, and when cross-compiling the archives should be cached already (as [termux_download](https://github.com/termux/termux-packages/blob/095442ec894c496f89db2127eefd8f6d0572412a/scripts/build/setup/termux_setup_golang.sh#L21) is used)

**Comment by thunder-coding** [comment]
I am talking about dependencies installed when running `go get` or `cargo build` while compiling packages. They aren't cached, and are downloaded again incase of rebuilding (only when the package is not installed the same version)

**Comment by Grimler91** [comment]
> I am talking about dependencies installed when running `go get` or `cargo build` while compiling packages. They aren't cached, and are downloaded again incase of rebuilding (only when the package is not installed the same version)

Right, those. That is a good suggestion, I am not very well-versed in the golang and rust build system though so not sure how difficult (or rather how many hacks) that would require.

I'll add it to my list of possible improvements (if someone else wants to look into it then that would be great)

**Review by viviskas — CHANGES_REQUESTED** [review]
#7108

**Review by csmart29 — COMMENTED** [review]
data/data/com.termux/files/usr

**Comment by Grimler91** [comment]
Let's close this old PR and re-open when it is closer to being ready. I'll periodically update the build-on-device branch.

**Pull request closed without merging**